tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -2577,6 +2599,200 @@ llvm::Optional<Value> convertReduceMeanOp(
   return val;
 }
 
+// Lowers ResizeBilinear and ResizeNearestNeighbor to TOSA resize.
+llvm::Optional<Value> convertResizeOp(PatternRewriter& rewriter, Operation* op,
+                                      RankedTensorType output_type,
+                                      Value input_value, StringRef mode,
+                                      bool align_corners,
+                                      bool half_pixel_centers) {
+  RankedTensorType input_type =
+      input_value.getType().dyn_cast<RankedTensorType>();
+  if (!input_type) return llvm::None;
+
+  if (input_type.getRank() != 4 || output_type.getRank() != 4) {
+    op->emitOpError("convertResizeOp: input/output must be rank 4");
+    return llvm::None;
+  }
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+
+  if (input_is_qtype != output_is_qtype) {
+    op->emitOpError(
+        "ConvertResizeOp: input/output tensor should "
+        "be all quantized or all floating-point.");
+    return llvm::None;
+  }
+
+  if (!input_is_qtype) {
+    if (!input_type.getElementType().isa<mlir::FloatType>()) {
+      op->emitOpError(
+          "ConvertResizeOp: only quantized or float types supported.");
+      return llvm::None;
+    }
+  }
+
+  auto input_shape = input_type.getShape();
+  auto output_shape = output_type.getShape();
+
+  size_t input_height = input_shape[1];
+  size_t input_width = input_shape[2];
+  size_t output_height = output_shape[1];
+  size_t output_width = output_shape[2];
+
+  // Need to resize around the center of the image to avoid distortion.
+  double in_center_h = static_cast<double>(input_height - 1) / 2.0;
+  double in_center_w = static_cast<double>(input_width - 1) / 2.0;
+  double out_center_h = static_cast<double>(output_height - 1) / 2.0;
+  double out_center_w = static_cast<double>(output_width - 1) / 2.0;
+
+  double fp_stride_y =
+      static_cast<double>(input_height) / static_cast<double>(output_height);
+  double fp_stride_x =
+      static_cast<double>(input_width) / static_cast<double>(output_width);
+  if (align_corners && output_height > 1) {
+    fp_stride_y = static_cast<double>(input_height - 1) /
+                  static_cast<double>(output_height - 1);
+  }
+  if (align_corners && output_width > 1) {
+    fp_stride_x = static_cast<double>(input_width - 1) /
+                  static_cast<double>(output_width - 1);
+  }
+
+  double fp_offset_y, fp_offset_x;
+  if (half_pixel_centers) {
+    fp_offset_y = fp_stride_y * 0.5f - 0.5f;
+    fp_offset_x = fp_stride_x * 0.5f - 0.5f;
+  } else {
+    fp_offset_y = 0.0f;
+    fp_offset_x = 0.0f;
+  }
+
+  // oh * fp_stride_y + fp_offset_y = ix
+
+  ArrayAttr output_size =
+      rewriter.getI64ArrayAttr({output_height, output_width});
```

Comment:
output_height and output_width require explicit static_casts for the initializer list.